File: orca/permissions.py
"""Permission engine — friction scaled to how dangerous an action is.

Modes:
  * default      — reads always allowed; edits ask once per session; bash asks
                   unless clearly read-only AND metacharacter-free, or allowlisted
  * acceptEdits  — file edits auto-approved; bash and web still gated
  * plan         — read-only exploration; writes, bash and web fetches are denied
  * yolo         — everything allowed except catastrophic commands

Rules look like "Bash(git *)" (fnmatch on the command), "Edit", "Write",
"read_file", "web_search" etc. Deny beats allow. Rules persist in
.orca/settings*.json.
"""
from __future__ import annotations

import fnmatch
import re
from typing import Any, Dict, List, Optional, Tuple

from .tools import CATASTROPHIC_BASH

MODES = ("default", "acceptEdits", "plan", "yolo")

# Commands auto-allowed in every mode *only when they contain no shell
# metacharacters* — `ls; rm -rf ~` must never slip through the prefix check.
SAFE_BASH_PREFIXES = (
    "ls", "pwd", "echo", "cat ", "head ", "tail ", "wc ", "file ",
    "git status", "git diff", "git log", "git show", "git branch", "git stash list",
    "grep ", "rg ", "which ", "whoami", "date", "uname",
    "python --version", "python3 --version", "node --version", "npm --version",
    "pip list", "pip freeze", "git --version", "docker ps", "docker images",
)

# Anything that can chain, redirect, substitute or execute another command.
_SHELL_META = re.compile(r"[;&|<>`\n\r]|\$\(|\|\||&&|\bexec\b|\bxargs\b")

PLAN_MODE_NOTICE = (
    "Plan mode is active: file modifications, command execution and web access "
    "are disabled. Explore the codebase (read_file/grep/glob/ls) and present a "
    "concrete implementation plan. The user will switch to execution mode after review."
)


class PermissionDenied(Exception):
    def __init__(self, message: str):
        super().__init__(message)
        self.message = message


def is_readonly_bash(command: str) -> bool:
    """True only for simple, metacharacter-free, known-read-only commands."""
    command = (command or "").strip()
    if not command or _SHELL_META.search(command):
        return False
    return any(command == p.rstrip() or command.startswith(p) for p in SAFE_BASH_PREFIXES)


class Permissions:
    def __init__(self, mode: str = "default", allow: Optional[List[str]] = None,
                 deny: Optional[List[str]] = None):
        if mode not in MODES:
            raise ValueError(f"Unknown permission mode '{mode}' (use {'/'.join(MODES)})")
        self.mode = mode
        self.allow: List[str] = list(allow or [])
        self.deny: List[str] = list(deny or [])

    # -- evaluation ----------------------------------------------------------

    def check(self, tool: str, args: Dict[str, Any], root=None) -> str:
        """Return 'allow', 'ask' or a denial reason string prefixed with 'deny:'."""
        perm = _perm_category(tool)

        if self._matches(self.deny, tool, args):
            return f"deny:Blocked by a deny rule for {tool}."

        if perm == "bash" and any(p.search(args.get("command", "")) for p in CATASTROPHIC_BASH):
            return "deny:Refused — this command looks catastrophic and is blocked in every mode."

        if self.mode == "yolo":
            return "allow"
        if self.mode == "plan":
            if perm in ("write", "bash", "web"):
                return f"deny:{PLAN_MODE_NOTICE}"
            return "allow"  # reads
        if self._matches(self.allow, tool, args):
            return "allow"
        if self.mode == "acceptEdits" and perm == "write":
            return "allow"

        if perm in ("read", "none"):
            return "allow"
        if perm == "bash":
            if is_readonly_bash(args.get("command", "")):
                return "allow"
            return "ask"
        if perm == "web":
            return "ask"
        # write in default mode: ask
        return "ask"

    # -- rules ---------------------------------------------------------------

    def allow_session(self, rule: str) -> None:
        if rule not in self.allow:
            self.allow.append(rule)

    # short names are accepted as aliases
    _ALIASES = {"edit": "edit_file", "write": "write_file", "read": "read_file",
                "webfetch": "web_fetch", "websearch": "web_search", "glob": "glob",
                "grep": "grep", "ls": "ls", "todo": "todo"}

    def _matches(self, rules: List[str], tool: str, args: Dict[str, Any]) -> bool:
        for rule in rules:
            if "(" in rule and rule.endswith(")"):
                name, _, pattern = rule[:-1].partition("(")
                if self._ALIASES.get(name.lower(), name.lower()) != tool:
                    continue
                if tool == "bash":
                    if fnmatch.fnmatch((args.get("command") or "").strip(), pattern):
                        return True
                elif tool in ("write_file", "edit_file"):
                    if fnmatch.fnmatch(args.get("path", ""), pattern):
                        return True
                else:
                    return True
            elif self._ALIASES.get(rule.lower(), rule.lower()) == tool:
                return True
        return False


def _perm_category(tool: str) -> str:
    from .tools import TOOL_BY_NAME
    spec = TOOL_BY_NAME.get(tool)
    return spec["perm"] if spec else "none"


def describe_tool_use(tool: str, args: Dict[str, Any], root=None) -> Tuple[str, str]:
    """(one-line label, multi-line detail) used in permission prompts & banners."""
    if tool == "bash":
        return "bash", args.get("command", "")
    if tool in ("write_file", "edit_file"):
        path = args.get("path", "")
        if tool == "write_file":
            return "write_file", f"path: {path}\n({len(args.get('content', ''))} chars of content)"
        return "edit_file", (f"path: {path}\nold: {args.get('old_string', '')[:200]!r}\n"
                             f"new: {args.get('new_string', '')[:200]!r}")
    if tool == "grep":
        return "grep", args.get("pattern", "")
    if tool in ("read_file", "glob", "ls"):
        primary = args.get("path") or args.get("pattern") or "."
        return tool, str(primary)
    if tool == "todo":
        items = args.get("todos") or []
        return "todo", f"{len(items)} items"
    if tool == "web_search":
        return "web_search", args.get("query", "")
    if tool == "web_fetch":
        return "web_fetch", args.get("url", "")
    return tool, str(args)[:400]

File: internal/permissions/permissions.go
// Package permissions decides allow / ask / deny for every tool call.
// Friction scales with danger; nothing dangerous is ever auto-approved
// based on regex blacklists alone.
package permissions

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Nethyric/orca/internal/security"
)

type Mode string

const (
	Default     Mode = "default"     // reads free; writes/bash/web ask (safe read-only bash auto)
	AcceptEdits Mode = "acceptEdits" // file edits auto; bash/web still gated
	Plan        Mode = "plan"        // read-only exploration
	Yolo        Mode = "yolo"        // everything allowed except catastrophic
)

var Modes = []Mode{Default, AcceptEdits, Plan, Yolo}

func ParseMode(s string) (Mode, error) {
	for _, m := range Modes {
		if string(m) == s {
			return m, nil
		}
	}
	return "", fmt.Errorf("unknown permission mode %q (default|acceptEdits|plan|yolo)", s)
}

type Verdict string

const (
	Allow Verdict = "allow"
	Ask   Verdict = "ask"
	Deny  Verdict = "deny"
)

type Decision struct {
	Verdict Verdict
	Reason  string
}

const PlanNotice = "Plan mode is active: file modification, command execution and web access are disabled. " +
	"Explore with read_file/grep/glob/ls and present a concrete plan."

type Permissions struct {
	Mode  Mode
	Allow []string
	Deny  []string
}

func New(mode Mode, allow, deny []string) *Permissions {
	return &Permissions{Mode: mode, Allow: allow, Deny: deny}
}

// Check evaluates one tool call. perm is the tool's category
// (read|write|bash|web|none), root the project root.
func (p *Permissions) Check(root, tool, perm string, args map[string]any) Decision {
	command, _ := args["command"].(string)

	if p.matches(p.Deny, tool, args) {
		return Decision{Deny, fmt.Sprintf("blocked by a deny rule for %s", tool)}
	}
	if perm == "bash" && security.IsCatastrophic(command) {
		return Decision{Deny, "refused — catastrophic command, blocked in every mode"}
	}
	if p.Mode == Yolo {
		return Decision{Allow, ""}
	}
	if p.Mode == Plan {
		if perm == "write" || perm == "bash" || perm == "web" {
			return Decision{Deny, PlanNotice}
		}
		return Decision{Allow, ""}
	}
	if p.matches(p.Allow, tool, args) {
		return Decision{Allow, "allow rule"}
	}
	if p.Mode == AcceptEdits && perm == "write" {
		return Decision{Allow, "acceptEdits"}
	}
	switch perm {
	case "read", "none":
		return Decision{Allow, ""}
	case "bash":
		if security.IsReadOnly(root, command) {
			return Decision{Allow, "read-only command"}
		}
		return Decision{Ask, ""}
	default: // write, web
		return Decision{Ask, ""}
	}
}

// AllowSession appends a session-scoped allow rule.
func (p *Permissions) AllowSession(rule string) {
	for _, r := range p.Allow {
		if r == rule {
			return
		}
	}
	p.Allow = append(p.Allow, rule)
}

var aliases = map[string]string{
	"edit": "edit_file", "write": "write_file", "read": "read_file",
	"webfetch": "web_fetch", "websearch": "web_search",
}

func canonical(name string) string {
	n := strings.ToLower(name)
	if a, ok := aliases[n]; ok {
		return a
	}
	return n
}

// matches evaluates rules like `bash(git *)`, `edit_file(src/*)`, `web_fetch`.
func (p *Permissions) matches(rules []string, tool string, args map[string]any) bool {
	for _, rule := range rules {
		if i := strings.IndexByte(rule, '('); i > 0 && strings.HasSuffix(rule, ")") {
			name, pattern := rule[:i], rule[i+1:len(rule)-1]
			if canonical(name) != tool {
				continue
			}
			switch tool {
			case "bash":
				cmd, _ := args["command"].(string)
				if globLike(pattern, strings.TrimSpace(cmd)) {
					return true
				}
			case "write_file", "edit_file":
				path, _ := args["path"].(string)
				if globLike(pattern, path) {
					return true
				}
			default:
				return true
			}
		} else if canonical(rule) == tool {
			return true
		}
	}
	return false
}

// globLike does shell-style matching where * spans any characters.
func globLike(pattern, s string) bool {
	var b strings.Builder
	b.WriteString("^")
	for _, r := range pattern {
		switch r {
		case '*':
			b.WriteString(".*")
		case '?':
			b.WriteString(".")
		default:
			b.WriteString(regexp.QuoteMeta(string(r)))
		}
	}
	b.WriteString("$")
	re, err := regexp.Compile(b.String())
	return err == nil && re.MatchString(s)
}
